packages/program/src/contract-call-script.ts

```diff
@@ -91,55 +120,112 @@ const getMainCallReceipt = (
+    }
 
-  const mainCallResult = getMainCallReceipt(
-    result.receipts as TransactionResultCallReceipt[],
-    contractId.toB256()
-  );
-  const mainCallInstructionStart = bn(mainCallResult?.is);
-  const receipts = result.receipts as ReturnReceipt[];
-  return receipts
-    .filter(({ type, is }) => isReturnType(type) && mainCallInstructionStart.eq(bn(is)))
-    .map((receipt: ReturnReceipt) => {
-      if (receipt.type === ReceiptType.Return) {
-        return new U64Coder().encode((receipt as TransactionResultReturnReceipt).val);
-      }
-      if (receipt.type === ReceiptType.ReturnData) {
-        const encodedScriptReturn = arrayify(receipt.data);
-        return encodedScriptReturn;
-      }
+    const mainCallResult = getMainCallReceipt(
+      result.receipts as TransactionResultCallReceipt[],
+      contractId.toB256()
+    );
+    const mainCallInstructionStart = bn(mainCallResult?.is);
 
-      return new Uint8Array();
-    });
-};
+    const receipts = result.receipts as ReturnReceipt[];
+    return receipts
+      .filter(({ type }) => isReturnType(type))
+      .flatMap((receipt: ReturnReceipt, index, filtered) => {
+        if (!mainCallInstructionStart.eq(bn(receipt.is))) {
+          return [];
+        }
+        if (receipt.type === ReceiptType.Return) {
+          return [new U64Coder().encode((receipt as TransactionResultReturnReceipt).val)];
+        }
+        if (receipt.type === ReceiptType.ReturnData) {
+          const encodedScriptReturn = arrayify(receipt.data);
+          if (isOutputDataHeap && isReturnType(filtered[index + 1]?.type)) {
+            const nextReturnData: TransactionResultReturnDataReceipt = filtered[
+              index + 1
+            ] as TransactionResultReturnDataReceipt;
+            return concat([encodedScriptReturn, arrayify(nextReturnData.data)]);
+          }
+
+          return [encodedScriptReturn];
+        }
+
+        return [new Uint8Array()];
+      })
+      .filter((v) => v !== undefined);
+  };
 
 export const decodeContractCallScriptResult = (
   callResult: CallResult,
   contractId: AbstractAddress,
+  isOutputDataHeap = false,
   logs: Array<any> = []
-): Uint8Array[] => decodeCallResult(callResult, scriptResultDecoder(contractId), logs);
+): Uint8Array[] =>
+  decodeCallResult(callResult, scriptResultDecoder(contractId, isOutputDataHeap), logs);
+
+const getCallInstructionsLength = (contractCalls: ContractCall[]): number => {
+  const singleStackCallLength = getSingleCallInstructions(
+    DEFAULT_OPCODE_PARAMS,
+    DEFAULT_OUTPUT_INFO
+  ).byteLength();
+
+  let totalHeapCalls = 0;
+  const heapCallsInstructionsLength = contractCalls.reduce((sum, call) => {
+    if (!call.isOutputDataHeap) {
+      return sum;
+    }
+
+    totalHeapCalls += 1;
+    return (
+      sum +
+      getSingleCallInstructions(DEFAULT_OPCODE_PARAMS, {
+        isHeap: true,
+        encodedLength: call.outputEncodedLength,
+      }).byteLength()
+    );
+  }, 0);
```

Comment:
```suggestion
  let totalHeapCalls = contractCalls.filter(call => call.isOutputDataHeap).length;
  const heapCallsInstructionsLength = contractCalls.reduce((sum, call) => {
    if (!call.isOutputDataHeap) {
      return sum;
    }

    return (
      sum +
      getSingleCallInstructions(DEFAULT_OPCODE_PARAMS, {
        isHeap: true,
        encodedLength: call.outputEncodedLength,
      }).byteLength()
    );
  }, 0);
```

This is better because it's not modifying an external variable (`totalHeapCalls`) from `.reduce()`, thus keeping it pure.